dev-tools/omdb/src/bin/omdb/db.rs

```diff
@@ -2249,6 +2416,1221 @@ async fn cmd_db_network_list_vnics(
+            row_char_counts.iter().map(|x| x.3).max().unwrap(),
+            "node id".len(),
+        ),
+        std::cmp::max(
+            row_char_counts.iter().map(|x| x.4).max().unwrap(),
+            "event type".len(),
+        ),
+    );
+
+    println!(
+        "{:>width0$} | {:width1$} | {:width2$} | {:width3$} | {:width4$} | {}",
+        String::from("saga id"),
+        String::from("event time"),
+        String::from("sub saga"),
+        String::from("node id"),
+        String::from("event type"),
+        String::from("data"),
+    );
+
+    println!(
+        "{:>width0$} | {:width1$} | {:width2$} | {:width3$} | {:width4$} | {}",
+        (0..width0).map(|_| "-").collect::<String>(),
+        (0..width1).map(|_| "-").collect::<String>(),
+        (0..width2).map(|_| "-").collect::<String>(),
+        (0..width3).map(|_| "-").collect::<String>(),
+        (0..width4).map(|_| "-").collect::<String>(),
+        String::from("---"),
+    );
+
+    for row in rows {
+        println!(
+            "{:>width0$} | {:width1$} | {:>width2$} | {:width3$} | {:width4$} | {}",
+            row.saga_id,
+            row.event_time,
+            if let Some(sub_saga_id) = row.sub_saga_id {
+                format!("{}", sub_saga_id)
+            } else {
+                String::from("")
+            },
+            row.node_id,
+            row.event_type,
+            row.data,
+        );
+    }
+}
+
+/// List all sagas
+async fn cmd_db_sagas_list(
+    datastore: &DataStore,
+    fetch_opts: &DbFetchOptions,
+    args: &SagaListArgs,
+) -> Result<(), anyhow::Error> {
+    let sagas = datastore
+        .pool_connection_for_tests()
+        .await?
+        .transaction_async(|conn| async move {
+            use db::schema::saga::dsl;
+
+            // Sorting by time_created requires a full table scan with the
+            // current index definitions.
+            conn.batch_execute_async(ALLOW_FULL_TABLE_SCAN_SQL).await?;
+
+            db::paginated(
+                dsl::saga,
+                dsl::time_created,
+                &first_page::<dsl::time_created>(fetch_opts.fetch_limit),
+            )
+            .load_async(&conn)
+            .await
+        })
+        .await?;
+
+    check_limit(&sagas, fetch_opts.fetch_limit, || {
+        String::from("listing sagas")
+    });
+
+    print_sagas(sagas, args.show_params);
+
+    Ok(())
+}
+
+/// List all stuck sagas
+async fn cmd_db_sagas_list_stuck(
+    datastore: &DataStore,
+    fetch_opts: &DbFetchOptions,
+) -> Result<(), anyhow::Error> {
+    let sagas = datastore
+        .pool_connection_for_tests()
+        .await?
+        .transaction_async(|conn| async move {
+            use db::schema::saga_node_event::dsl;
+
+            // Sorting by time_created requires a full table scan with the
+            // current index definitions.
+            conn.batch_execute_async(ALLOW_FULL_TABLE_SCAN_SQL).await?;
+
+            let stuck_sagas: Vec<SagaId> = db::paginated(
```

Comment:
As above: if you're going to get all the results, you want to use `Paginator` I think.  If not, you don't need `db::paginated` (but then you should probably use `check_limit` to warn the user if your results are truncated).

You could also do both of these in one query (something like `SELECT saga_id from saga s, saga_node_event sne WHERE s.id = sne.saga_id AND sne.event_type == 'undo_failed').  I think this should be possible with regular diesel but may be a pain and definitely not necessary here.

It seems like we use this pattern in a few places (fetch saga ids (or maybe even entire records?) for sagas with a node event of type `type`).  It might help to make a helper function.